pkg/intermediate/aggregate.go

```diff
@@ -215,6 +215,71 @@ func (a *AggregationProcess) GetExpiryFromExpirePriorityQueue() time.Duration {
+		if flowKey != nil {
+			if (flowKey.SourceAddress != "" && flowKey.SourceAddress != currentFlowKey.SourceAddress) ||
+				(flowKey.DestinationAddress != "" && flowKey.DestinationAddress != currentFlowKey.DestinationAddress) ||
+				(flowKey.Protocol != 0 && flowKey.Protocol != currentFlowKey.Protocol) ||
+				(flowKey.SourcePort != 0 && flowKey.SourcePort != currentFlowKey.SourcePort) ||
+				(flowKey.DestinationPort != 0 && flowKey.DestinationPort != currentFlowKey.DestinationPort) {
+				continue
+			}
+		}
+		var recordString string
+		for _, ie := range record.Record.GetOrderedElementList() {
+			switch ie.GetDataType() {
+			case entities.Unsigned8:
+				recordString += fmt.Sprintf("    %s: %v \n", ie.GetName(), ie.GetUnsigned8Value())
+			case entities.Unsigned16:
+				recordString += fmt.Sprintf("    %s: %v \n", ie.GetName(), ie.GetUnsigned16Value())
+			case entities.Unsigned32:
+				recordString += fmt.Sprintf("    %s: %v \n", ie.GetName(), ie.GetUnsigned32Value())
+			case entities.Unsigned64:
+				recordString += fmt.Sprintf("    %s: %v \n", ie.GetName(), ie.GetUnsigned64Value())
+			case entities.Signed8:
+				recordString += fmt.Sprintf("    %s: %v \n", ie.GetName(), ie.GetSigned8Value())
+			case entities.Signed16:
+				recordString += fmt.Sprintf("    %s: %v \n", ie.GetName(), ie.GetSigned16Value())
+			case entities.Signed32:
+				recordString += fmt.Sprintf("    %s: %v \n", ie.GetName(), ie.GetSigned32Value())
+			case entities.Signed64:
+				recordString += fmt.Sprintf("    %s: %v \n", ie.GetName(), ie.GetSigned64Value())
+			case entities.Float32:
+				recordString += fmt.Sprintf("    %s: %v \n", ie.GetName(), ie.GetFloat32Value())
+			case entities.Float64:
+				recordString += fmt.Sprintf("    %s: %v \n", ie.GetName(), ie.GetFloat64Value())
+			case entities.Boolean:
+				recordString += fmt.Sprintf("    %s: %v \n", ie.GetName(), ie.GetBooleanValue())
+			case entities.DateTimeSeconds:
+				recordString += fmt.Sprintf("    %s: %v \n", ie.GetName(), ie.GetUnsigned32Value())
+			case entities.DateTimeMilliseconds:
+				recordString += fmt.Sprintf("    %s: %v \n", ie.GetName(), ie.GetUnsigned64Value())
+			case entities.DateTimeMicroseconds, entities.DateTimeNanoseconds:
+				err := fmt.Errorf("API does not support micro and nano seconds types yet")
+				recordString += fmt.Sprintf("    %s: %v \n", ie.GetName(), err)
+			case entities.MacAddress:
+				recordString += fmt.Sprintf("    %s: %v \n", ie.GetName(), ie.GetMacAddressValue())
+			case entities.Ipv4Address, entities.Ipv6Address:
+				recordString += fmt.Sprintf("    %s: %v \n", ie.GetName(), ie.GetIPAddressValue())
+			case entities.String:
+				recordString += fmt.Sprintf("    %s: %v \n", ie.GetName(), ie.GetStringValue())
+			default:
+				err := fmt.Errorf("API supports only valid information elements with datatypes given in RFC7011")
+				recordString += fmt.Sprintf("    %s: %v \n", ie.GetName(), err)
+			}
+		}
+		records = append(records, recordString)
```

Comment:
As there is only one unique flow key in the map, if the flowKey is not nil, then we can break after this point.

Reply:
Would add a hash check before the loop. Here I would like to add some flexibility in the search. For example, if the user only provides the source address, all the flow records with that source address would be returned. In this case, it seems we cannot add the break here.

Reply:
Ok. If that is the case, we can break it into two scenarios: complete filter and partial filter.
In the case of complete filter, you can directly check in the map. And with partial filter, you can use your current approach.

In the future, we can explore moving to `cache.Indexer` from the [kubernetes client-go package](https://github.com/kubernetes/client-go/blob/v0.22.2/tools/cache/index.go#L35). This makes partial filter search efficient.

Reply:
Updated